Add scala tests to CI

johnynek opened the pull request:
This needs to be after #862

johnynek commented:
@damienmg master does not reflect the change i #862 so, this does not pass until it does.

johnynek commented:
@damienmg is anything needed for this?

damienmg commented:
LGTM I will merge it tomorrow

Thanks again!

kchodorow commented on tools/build_defs/scala/scala.bzl:
```diff
@@ -121,14 +121,22 @@ def _collect_comp_run_jars(ctx):
       compile_jars += target.java.transitive_deps
+      found = True
+    if not found:
+       #support http_file pointed at a jar. http_jar uses ijar, which breaks scala macros
```
Please change to "# support" (1 space between # and the comment).

kchodorow commented on compile.sh:
```diff
@@ -154,6 +154,7 @@ if [ $DO_TESTS ]; then
       ${EXTRA_BAZEL_ARGS} \
       --javacopt="-source ${JAVA_VERSION} -target ${JAVA_VERSION}" \
       -k --test_output=errors //src/... //third_party/ijar/... //scripts/... \
+      tools/build_defs/scala/test/... \
```
//tools/build_defs/scala/test/...

johnynek commented:
Something is broken in bazel 0.1.5 with this. Running the tests I get all kinds of errors related to other skylark builds:
```
st-oscar1:bazel oscar$ bazel test //tools/build_defs/scala/test/...
.....
ERROR: /Users/oscar/oss/bazel/tools/build_defs/dotnet/csharp.bzl:306:3: //external:nunit: no such attribute 'build_file_co
ntent' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/dotnet/csharp.bzl:306:3: //external:nunit: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/d/d.bzl:507:3: //external:dmd_linux_x86_64: no such attribute 'build_file_content' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/d/d.bzl:507:3: //external:dmd_linux_x86_64: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/d/d.bzl:514:3: //external:dmd_darwin_x86_64: no such attribute 'build_file_content' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/d/d.bzl:514:3: //external:dmd_darwin_x86_64: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_rules/go/def.bzl:399:3: //external:golang_linux_amd64: no such attribute 'build_file_content' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_rules/go/def.bzl:399:3: //external:golang_linux_amd64: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_rules/go/def.bzl:406:3: //external:golang_darwin_amd64: no such attribute 'build_file_content' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_rules/go/def.bzl:406:3: //external:golang_darwin_amd64: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_rules/rust/rust.bzl:739:3: //external:rust_linux_x86_64: no such attribute 'build_file_content' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_rules/rust/rust.bzl:739:3: //external:rust_linux_x86_64: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_rules/rust/rust.bzl:747:3: //external:rust_darwin_x86_64: no such attribute 'build_file_content' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_rules/rust/rust.bzl:747:3: //external:rust_darwin_x86_64: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/sass/sass.bzl:127:3: //external:libsass: no such attribute 'build_file_content' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/sass/sass.bzl:127:3: //external:libsass: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/sass/sass.bzl:135:3: //external:sassc: no such attribute 'build_file_content' in 'new_http_archive' rule.
ERROR: /Users/oscar/oss/bazel/tools/build_defs/sass/sass.bzl:135:3: //external:sassc: missing value for mandatory attribute 'build_file' in 'new_http_archive' rule.
ERROR: package contains errors: tools/build_defs/scala/test.
ERROR: no such package 'external': Package 'external' contains errors.
INFO: Elapsed time: 0.406s
ERROR: Couldn't start the build. Unable to run tests.
```
Why should errors in other files break this test? For instance:
https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/sass/sass.bzl#L135

does have 'build_file_content' and not a build_file. Is that wrong now?
I wonder if since the other rules aren't in CI they are broken. Any idea what to try?

johnynek commented:
I guess actually it is the opposite, the code now exercises features not in 0.1.5.

Okay, I addressed the comments and also fixed another issue. I made this more similar to the java rule. I fixed an issue that compilation deps were transitive (unlike java which follows a better practice of requiring declarations).

One issue: I don't see how to access `exports` in a java rule. I work around this now by using transitive java compile deps, which I assume includes the exports. If there is a way to access that list. I will use it.

This code also builds a small repo of scala code I have.

kchodorow commented:
Yeah, I think this is an incompatibility between Bazel at 0.1.5 and the WORKSPACE rules at HEAD.  175b267569219100ac756b0ef016211b0cfa90c4 added the build_file_content option for new_ repos, and we quickly switched all the built-in Skylark rules over.

I don't know about the exports thing, maybe @damienmg can help with that.

damienmg commented:
FYI, this has been merged.

Already answered on the ML for exports thing. And for the
build_file_content, we moved to it because build_file cannot refer to an
external repository (yet?).

On Fri, Feb 12, 2016 at 8:34 PM Kristina <notifications@github.com> wrote:

> Yeah, I think this is an incompatibility between Bazel at 0.1.5 and the
> WORKSPACE rules at HEAD. 175b267
> <https://github.com/bazelbuild/bazel/commit/175b267569219100ac756b0ef016211b0cfa90c4>
> added the build_file_content option for new_ repos, and we quickly switched
> all the built-in Skylark rules over.
>
> I don't know about the exports thing, maybe @damienmg
> <https://github.com/damienmg> can help with that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bazelbuild/bazel/pull/872#issuecomment-183459860>.
>

johnynek commented:
Thanks!

damienmg commented:
Merge got in!

johnynek commented:
@damienmg Thank you, sir!
